addons/udes_stock/models/stock_move_line.py

```diff
@@ -252,3 +252,132 @@ def get_move_lines_ordered_by(self, domain, aux_domain=None, order="id"):
         if aux_domain is not None:
             domain += aux_domain
         return StockMoveLine.search(domain, order=order)
+
+    def _find_move_lines(self, uom_qty, product=None, package=None, lot_name=None, location=None):
+        """ Find a subset of move lines from self matching the subset of the key:
+            product, package, lot and location.
+        """
+        funcs = []
+        if product:
+            funcs.append(lambda ml: ml.product_id == product)
+        if package:
+            funcs.append(lambda ml: ml.package_id == package)
+        if location:
+            funcs.append(lambda ml: ml.location_id == location)
+        if lot_name:
+            funcs.append(lambda ml: (lot_name == lot_name or ml.lot_id.name == lot_name))
+        func = lambda ml: all(f(ml) for f in funcs)
+
+        # Find move lines
+        mls = self.filtered(func)
+        mls_fulfill, new_ml, uom_qty = mls.move_lines_for_qty(uom_qty)
+        if uom_qty > 0:
+            self.picking_id.add_unexpected_parts()
+        return mls_fulfill, new_ml
+
+    def prepare_move_lines(
+        self, product_ids=None, package=None, location=None, result_package=None, location_dest=None
+    ):
+        """ Prepare incomplete move lines in self into groups of mls to be updated with
+            the same values.
+        """
+        Product = self.env["product.product"]
+
+        vals = {}
+        if result_package:
+            vals["result_package_id"] = result_package.id
+        if location_dest:
+            vals["location_dest_id"] = location_dest.id
+        mls = self.get_lines_incomplete()
+        incoming = mls.picking_type_id.code == "incoming"
+        res = defaultdict(dict)
+        if product_ids:
```

Comment:
I would suggest to delete if product_ids and loop directly through them.
After the loop 
at the beginning of the line smth like:

```
if not product_ids:
    if package:
        pack_mls = package._get_current_move_lines() ....
   else:
         res[mls] = vals
```

The reason why i am suggesting this is because you gain some characters on the product_ids loop and i think looks better the code